includes/class-amp-story-post-type.php

```diff
@@ -1553,4 +1589,222 @@ public static function add_new_max_image_size( $image_sizes ) {
+			],
+			500
+		);
+	}
+
+	/**
+	 * Generates a Zip archive from the AMP Story.
+	 *
+	 * @param int $post_id The post ID of the AMP Story.
+	 * @return WP_Error
+	 */
+	private static function generate_export( $post_id ) {
+		$post = get_post( $post_id );
+
+		if ( ! $post ) {
+			return new WP_Error( 'amp_story_export_invalid_post', esc_html__( 'The AMP Story does not exist.', 'amp' ) );
+		}
+
+		$slug = sanitize_title( $post->post_title, $post->ID );
+		$file = wp_tempnam( $slug );
+
+		$zip = new ZipArchive();
+		$res = $zip->open( $file, ZipArchive::CREATE | ZipArchive::OVERWRITE );
+
+		if ( true !== $res ) {
+			/* translators: %s is the ZipArchive error code. */
+			return new WP_Error( 'amp_story_zip_archive_error', sprintf( esc_html__( 'There was an error generating the ZIP archive. Error code: %s', 'amp' ), $res ) );
+		}
+
+		// Passed to `get_preview_post_link()` for nonce access and to sanitize the output.
+		$query_args = [
+			'story_export' => true,
+			'_wpnonce'     => wp_create_nonce( self::AMP_STORIES_AJAX_ACTION ),
+		];
+
+		// Passed to `wp_remote_get()`.
+		$args = [
+			'cookies'     => wp_unslash( $_COOKIE ), // Pass along cookies so private pages and drafts can be accessed.
+			'timeout'     => 15, // Increase from default of 5 to give extra time for the plugin to identify the sources for any given validation errors.
+			'sslverify'   => false,
+			'redirection' => 0, // Because we're in a loop for redirection.
+			'headers'     => [
+				'Cache-Control' => 'no-cache',
+			],
+		];
+
+		// Get the preview URL.
+		$response = wp_remote_get( get_preview_post_link( $post, $query_args ), $args );
+
+		// Ensure we have the required data.
+		if ( ! ( is_array( $response ) && isset( $response['body'] ) ) ) {
+			return new WP_Error( 'amp_story_export_response', esc_html__( 'Could not get the AMP Story HTML.', 'amp' ) );
+		}
+
+		// Get the HTML from the response body.
+		$html   = $response['body'];
+		$assets = [];
+		$regex  = '<!--\s*AMP_EXPORT_ASSETS\s*:\s*(\[.*?\])\s*-->';
+
+		// Get the assets from the AMP_EXPORT_ASSETS comment.
+		if ( preg_match( '#</body>.*?' . $regex . '#s', $html, $matches ) ) {
+			$assets = json_decode( $matches[1], true );
+
+			// Remove the comment.
+			$html = preg_replace( '/' . $regex . '/s', '', $html );
+		}
+
+		// Create the zip directory.
+		$zip->addEmptyDir( $slug );
+
+		// Add README.txt file.
+		$zip->addFromString( $slug . '/README.txt', 'temporary content...' );
+
+		// Add index.html file.
+		$zip->addFromString( $slug . '/index.html', $html );
+
+		// Add the assets.
+		if ( ! empty( $assets ) ) {
+
+			// Create the empty assets directory.
+			$zip->addEmptyDir( $slug . '/assets' );
+
+			foreach ( $assets as $asset ) {
+				$response = wp_remote_get( $asset );
+				if ( is_array( $response ) && ! empty( $response['body'] ) ) {
+					$zip->addFromString( $slug . '/assets/' . self::export_image_basename( $asset ), $response['body'] );
```

Comment:
Perhaps it would make sense to add an additional notice if some of the assets failed to load? E.g. if `$response` is an error, display a message to express that some of the assets failed to load; and potentially logging the errors somewhere, too. Would avoid debugging / support cases in case of server issues. It could fail due to SSL issues, timing out, etc. Thoughts?

cc @westonruter

Reply:
As for SSL issues, I would probably just disable SSL verification for the `wp_remote_get` call.

Reply:
If we return an error at this stage the whole process fails. It's kind of an all or nothing scenario.

Reply:
I've updated the code to address the SSL issue.